### PR TITLE
fix(typing): allow `Bot(help_command=None)`

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Fixes the `help_command` parameter annotations to allow `Optional[HelpCommand]` instead of `HelpCommand`, which the implementation already supports.

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [x] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/ext/commands/context.py [hunk 1]
```diff
@@ -324,7 +324,7 @@ async def send_help(self, *args: Any) -> Any:
         from .core import Command, Group, wrap_callback
         from .errors import CommandError
 
-        bot = self.bot
+        bot: Union[Bot, AutoShardedBot] = self.bot
```
Comment (onerandomusername) [inline disnake/ext/commands/context.py:327]: sneaky change

Comment (shiftinv) [reply disnake/ext/commands/context.py]: I was looking for occurrences of `help_command` and noticed that this entire method was mostly `Unknown`, but apparently pylance had just given up ~~for the umpteenth time today~~. Reverted in https://github.com/DisnakeDev/disnake/pull/849/commits/7b53fb022ddf31956867a9cdc190e6f0749b7ea2

File: disnake/ext/commands/bot_base.py [hunk 2]
```diff
@@ -433,18 +433,16 @@ def help_command(self) -> Optional[HelpCommand]:
 
     @help_command.setter
     def help_command(self, value: Optional[HelpCommand]) -> None:
+        if value is not None and not isinstance(value, HelpCommand):
```
Comment (onerandomusername) [inline disnake/ext/commands/bot_base.py:436]: why is this changed?

Comment (shiftinv) [reply disnake/ext/commands/bot_base.py:436]: Cleaning up some slightly convoluted logic, which was also partially the cause for a pyright bug with inferring the attribute type correctly; the runtime behavior should be unchanged.

**Review by onerandomusername — APPROVED** [review]
lgtm
